File: backend/sources/tank01.py
"""
Tank01 NFL API data source (RapidAPI)
API: https://rapidapi.com/tank01/api/nfl-live-in-game-real-time-statistics-nfl

Endpoints utilizados:
- /getNFLTeamRoster?teamAbv=XXX&getStats=true - Roster com stats
"""

import httpx
from typing import Optional
import sys
from pathlib import Path

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent))
from config import RAPIDAPI_KEY, TANK01_BASE_URL, REQUEST_TIMEOUT
from cache import read_cache, write_cache

# Todos os times NFL
NFL_TEAMS = [
    "ARI", "ATL", "BAL", "BUF", "CAR", "CHI", "CIN", "CLE",
    "DAL", "DEN", "DET", "GB", "HOU", "IND", "JAX", "KC",
    "LAC", "LAR", "LV", "MIA", "MIN", "NE", "NO", "NYG",
    "NYJ", "PHI", "PIT", "SF", "SEA", "TB", "TEN", "WAS"
]

# Mapeamento de posições para Fantasy
POSITION_MAP = {
    # DL (Defensive Line)
    "DE": "DL", "DT": "DL", "NT": "DL",
    # LB (Linebacker)
    "LB": "LB", "ILB": "LB", "OLB": "LB", "MLB": "LB",
    # DB (Defensive Back)
    "CB": "DB", "S": "DB", "FS": "DB", "SS": "DB", "DB": "DB",
}

DEFENSIVE_POSITIONS = list(POSITION_MAP.keys())
OFFENSIVE_POSITIONS = ["QB", "RB", "WR", "TE", "FB"]


def is_configured() -> bool:
    """Verifica se a API está configurada"""
    return bool(RAPIDAPI_KEY)


def get_headers() -> dict:
    """Retorna headers para requisições à API"""
    return {
        "X-RapidAPI-Key": RAPIDAPI_KEY,
        "X-RapidAPI-Host": "tank01-nfl-live-in-game-real-time-statistics-nfl.p.rapidapi.com"
    }


def safe_int(value, default=0) -> int:
    """Converte valor para int de forma segura"""
    if value is None:
        return default
    try:
        return int(float(str(value).replace(",", "")))
    except (ValueError, TypeError):
        return default


def safe_float(value, default=0.0) -> float:
    """Converte valor para float de forma segura"""
    if value is None:
        return default
    try:
        f = float(str(value).replace(",", ""))
        if f != f or f == float('inf') or f == float('-inf'):  # NaN or Infinity
            return default
        return f
    except (ValueError, TypeError):
        return default


async def fetch_team_roster_with_stats(team: str) -> list[dict]:
    """
    Busca roster de um time com stats
    Endpoint: /getNFLTeamRoster?teamAbv=XXX&getStats=true
    """
    if not is_configured():
        raise ValueError("Tank01 API not configured")

    url = f"{TANK01_BASE_URL}/getNFLTeamRoster"

    async with httpx.AsyncClient() as client:
        response = await client.get(
            url,
            headers=get_headers(),
            params={"teamAbv": team, "getStats": "true"},
            timeout=REQUEST_TIMEOUT
        )
        response.raise_for_status()
        data = response.json()
        return data.get("body", {}).get("roster", [])


async def fetch_all_players_with_stats() -> list[dict]:
    """
    Busca todos os jogadores de todos os times com stats
    Faz 32 chamadas (1 por time)
    """
    cache_key = "tank01_all_players"
    cached = read_cache(cache_key, 0)
    if cached is not None:
        return cached

    all_players = []

    for team in NFL_TEAMS:
        try:
            print(f"[tank01] Buscando roster de {team}...")
            roster = await fetch_team_roster_with_stats(team)
            all_players.extend(roster)
        except Exception as e:
            print(f"[tank01] Erro ao buscar {team}: {e}")

    if all_players:
        write_cache(cache_key, 0, all_players)

    return all_players


async def get_defensive_stats(season: int = 2024) -> list[dict]:
    """
    Retorna stats defensivas usando Tank01 API
    """
    cache_key = f"tank01_def_stats_{season}"
    cached = read_cache(cache_key, 0)
    if cached is not None:
        return cached

    if not is_configured():
        raise ValueError("Tank01 API not configured. Set RAPIDAPI_KEY environment variable.")

    # Busca todos os jogadores com stats
    players_raw = await fetch_all_players_with_stats()

    result = []

    for player_data in players_raw:
        pos = str(player_data.get("pos", "")).upper()

        # Só inclui posições defensivas
        if pos not in DEFENSIVE_POSITIONS:
            continue

        fantasy_pos = POSITION_MAP.get(pos, "LB")

        # Extrai stats do objeto - Tank01 usa nested structure
        stats_obj = player_data.get("stats", {}) or {}
        def_stats = stats_obj.get("Defense", {}) or {}

        tackles = safe_int(def_stats.get("totalTackles", 0))
        solo = safe_int(def_stats.get("soloTackles", 0))
        sacks = safe_float(def_stats.get("sacks", 0))
        tfl = safe_int(def_stats.get("tfl", 0))
        qb_hits = safe_int(def_stats.get("qbHits", 0))
        pd = safe_int(def_stats.get("passDeflections", 0))
        ints = safe_int(def_stats.get("defensiveInterceptions", 0))
        ff = safe_int(def_stats.get("forcedFumbles", 0))
        fr = safe_int(def_stats.get("fumblesRecovered", 0))
        def_td = safe_int(def_stats.get("defTD", 0))

        # Só adiciona se tiver stats relevantes
        if tackles == 0 and sacks == 0 and ints == 0:
            continue

        player = {
            "id": str(player_data.get("playerID", player_data.get("espnID", ""))),
            "name": player_data.get("longName") or player_data.get("espnName", "Unknown"),
            "team": player_data.get("team", ""),
            "teamAbbr": player_data.get("team", ""),
            "photoUrl": player_data.get("espnHeadshot", ""),
            "espnPosition": pos,
            "fantasyPosition": fantasy_pos,
            "age": safe_int(player_data.get("age")) if player_data.get("age") else None,
            "experience": safe_int(player_data.get("exp")) if player_data.get("exp") else None,
            "jerseyNumber": str(player_data.get("jerseyNum", "")) if player_data.get("jerseyNum") else None,
            "stats": {
                "tackles": tackles,
                "sacks": sacks,
                "tfl": tfl,
                "qbHits": qb_hits,
                "passesDefended": pd,
                "interceptions": ints,
                "forcedFumbles": ff,
                "soloTackles": solo,
                "assistTackles": tackles - solo if tackles > solo else 0,
                "tacklesWithAssist": 0,
                "tflYards": 0,
                "sackYards": 0,
                "intYards": 0,
                "defensiveTds": def_td,
                "fumbleRecoveryOwn": 0,
                "fumbleRecoveryOpp": fr,
                "safeties": 0,
            }
        }

        result.append(player)

    # Ordena por tackles (desc)
    result.sort(key=lambda x: x["stats"]["tackles"], reverse=True)

    # Cache o resultado
    if result:
        write_cache(cache_key, 0, result)

    return result


async def get_offensive_stats(season: int = 2024) -> list[dict]:
    """
    Retorna stats ofensivas usando Tank01 API
    """
    cache_key = f"tank01_off_stats_{season}"
    cached = read_cache(cache_key, 0)
    if cached is not None:
        return cached

    if not is_configured():
        raise ValueError("Tank01 API not configured. Set RAPIDAPI_KEY environment variable.")

    # Busca todos os jogadores com stats
    players_raw = await fetch_all_players_with_stats()

    result = []

    for player_data in players_raw:
        pos = str(player_data.get("pos", "")).upper()

        # Só inclui posições ofensivas
        if pos not in OFFENSIVE_POSITIONS:
            continue

        # Extrai stats
        stats_obj = player_data.get("stats", {}) or {}
        passing = stats_obj.get("Passing", {}) or {}
        rushing = stats_obj.get("Rushing", {}) or {}
        receiving = stats_obj.get("Receiving", {}) or {}
        defense = stats_obj.get("Defense", {}) or {}  # Para fumbles

        # Passing stats
        completions = safe_int(passing.get("passCompletions", 0))
        attempts = safe_int(passing.get("passAttempts", 0))
        passing_yards = safe_int(passing.get("passYds", 0))
        passing_tds = safe_int(passing.get("passTD", 0))
        ints = safe_int(passing.get("int", 0))
        sacks_taken = safe_int(passing.get("sacked", 0))

        # Rushing stats
        carries = safe_int(rushing.get("carries", 0))
        rushing_yards = safe_int(rushing.get("rushYds", 0))
        rushing_tds = safe_int(rushing.get("rushTD", 0))

        # Receiving stats
        targets = safe_int(receiving.get("targets", 0))
        receptions = safe_int(receiving.get("receptions", 0))
        receiving_yards = safe_int(receiving.get("recYds", 0))
        receiving_tds = safe_int(receiving.get("recTD", 0))

        # Fumbles
        fumbles = safe_int(defense.get("fumbles", 0))
        fumbles_lost = safe_int(defense.get("fumblesLost", 0))

        # Calcula passer rating
        passer_rating = 0.0
        if attempts > 0:
            a = max(0, min(2.375, ((completions / attempts) - 0.3) * 5))
            b = max(0, min(2.375, ((passing_yards / attempts) - 3) * 0.25))
            c = max(0, min(2.375, (passing_tds / attempts) * 20))
            d = max(0, min(2.375, 2.375 - ((ints / attempts) * 25)))
            passer_rating = round(((a + b + c + d) / 6) * 100, 1)

        total_yards = passing_yards + rushing_yards + receiving_yards
        if total_yards == 0:
            continue

        player = {
            "id": str(player_data.get("playerID", player_data.get("espnID", ""))),
            "name": player_data.get("longName") or player_data.get("espnName", "Unknown"),
            "team": player_data.get("team", ""),
            "teamAbbr": player_data.get("team", ""),
            "photoUrl": player_data.get("espnHeadshot", ""),
            "position": pos,
            "age": safe_int(player_data.get("age")) if player_data.get("age") else None,
            "experience": safe_int(player_data.get("exp")) if player_data.get("exp") else None,
            "jerseyNumber": str(player_data.get("jerseyNum", "")) if player_data.get("jerseyNum") else None,
            "stats": {
                # Passing
                "completions": completions,
                "attempts": attempts,
                "passingYards": passing_yards,
                "passingTds": passing_tds,
                "interceptions": ints,
                "passerRating": passer_rating,
                "passingAirYards": 0,
                "passingYac": 0,
                "passingFirstDowns": 0,
                "sacks": sacks_taken,
                "sackYards": 0,

                # Rushing
                "carries": carries,
                "rushingYards": rushing_yards,
                "rushingTds": rushing_tds,
                "rushingFirstDowns": 0,

                # Receiving
                "targets": targets,
                "receptions": receptions,
                "receivingYards": receiving_yards,
                "receivingTds": receiving_tds,
                "receivingAirYards": 0,
                "receivingYac": 0,
                "receivingFirstDowns": 0,

                # Fumbles
                "fumbles": fumbles,
                "fumblesLost": fumbles_lost,

                # Advanced (not available from Tank01 basic)
                "targetShare": 0,
                "airYardsShare": 0,
                "wopr": 0,
                "racr": 0,
                "pacr": 0,

                # Fantasy (calculated later if needed)
                "fantasyPoints": 0,
                "fantasyPointsPpr": 0,
            }
        }

        result.append(player)

    # Ordena por total yards
    result.sort(key=lambda x: x["stats"]["passingYards"] + x["stats"]["rushingYards"] + x["stats"]["receivingYards"], reverse=True)

    # Cache o resultado
    if result:
        write_cache(cache_key, 0, result)

    return result

File: workers/nfl-stats-api/src/sources/tank01.ts
/**
 * Tank01 NFL API data source (RapidAPI)
 * API: https://rapidapi.com/tank01/api/nfl-live-in-game-real-time-statistics-nfl
 */

import type { Env, Tank01Player, DefensivePlayer, OffensivePlayer } from '../types';

const TANK01_BASE_URL = 'https://tank01-nfl-live-in-game-real-time-statistics-nfl.p.rapidapi.com';

// All NFL teams
const NFL_TEAMS = [
  'ARI', 'ATL', 'BAL', 'BUF', 'CAR', 'CHI', 'CIN', 'CLE',
  'DAL', 'DEN', 'DET', 'GB', 'HOU', 'IND', 'JAX', 'KC',
  'LAC', 'LAR', 'LV', 'MIA', 'MIN', 'NE', 'NO', 'NYG',
  'NYJ', 'PHI', 'PIT', 'SF', 'SEA', 'TB', 'TEN', 'WAS',
];

// Position mappings for Fantasy
const POSITION_MAP: Record<string, string> = {
  // DL (Defensive Line)
  DE: 'DL', DT: 'DL', NT: 'DL',
  // LB (Linebacker)
  LB: 'LB', ILB: 'LB', OLB: 'LB', MLB: 'LB',
  // DB (Defensive Back)
  CB: 'DB', S: 'DB', FS: 'DB', SS: 'DB', DB: 'DB',
};

const DEFENSIVE_POSITIONS = Object.keys(POSITION_MAP);
const OFFENSIVE_POSITIONS = ['QB', 'RB', 'WR', 'TE', 'FB'];

/**
 * Check if Tank01 API is configured
 */
export function isConfigured(env: Env): boolean {
  return Boolean(env.RAPIDAPI_KEY);
}

/**
 * Get headers for Tank01 API requests
 */
function getHeaders(env: Env): HeadersInit {
  return {
    'X-RapidAPI-Key': env.RAPIDAPI_KEY,
    'X-RapidAPI-Host': 'tank01-nfl-live-in-game-real-time-statistics-nfl.p.rapidapi.com',
  };
}

/**
 * Safe integer conversion
 */
function safeInt(value: unknown, defaultValue = 0): number {
  if (value === null || value === undefined) return defaultValue;
  try {
    const num = parseInt(String(value).replace(',', ''), 10);
    return isNaN(num) ? defaultValue : num;
  } catch {
    return defaultValue;
  }
}

/**
 * Safe float conversion
 */
function safeFloat(value: unknown, defaultValue = 0): number {
  if (value === null || value === undefined) return defaultValue;
  try {
    const num = parseFloat(String(value).replace(',', ''));
    if (isNaN(num) || !isFinite(num)) return defaultValue;
    return num;
  } catch {
    return defaultValue;
  }
}

/**
 * Fetch roster for a single team with stats
 */
async function fetchTeamRoster(team: string, env: Env): Promise<Tank01Player[]> {
  const url = `${TANK01_BASE_URL}/getNFLTeamRoster?teamAbv=${team}&getStats=true`;

  const response = await fetch(url, {
    headers: getHeaders(env),
  });

  if (!response.ok) {
    throw new Error(`Failed to fetch roster for ${team}: ${response.status}`);
  }

  const data = await response.json() as { body?: { roster?: Tank01Player[] } };
  return data.body?.roster || [];
}

/**
 * Fetch all players from all teams with stats
 */
async function fetchAllPlayers(env: Env, cache: KVNamespace): Promise<Tank01Player[]> {
  const cacheKey = 'tank01_all_players';

  // Check cache first
  const cached = await cache.get(cacheKey);
  if (cached) {
    return JSON.parse(cached);
  }

  // Fetch from all teams in parallel (batched to avoid rate limits)
  const allPlayers: Tank01Player[] = [];
  const batchSize = 8;

  for (let i = 0; i < NFL_TEAMS.length; i += batchSize) {
    const batch = NFL_TEAMS.slice(i, i + batchSize);
    const results = await Promise.allSettled(
      batch.map(team => fetchTeamRoster(team, env))
    );

    for (const result of results) {
      if (result.status === 'fulfilled') {
        allPlayers.push(...result.value);
      }
    }
  }

  // Cache for 1 hour
  if (allPlayers.length > 0) {
    await cache.put(cacheKey, JSON.stringify(allPlayers), {
      expirationTtl: 3600,
    });
  }

  return allPlayers;
}

/**
 * Get defensive stats from Tank01 API
 */
export async function getDefensiveStats(
  season: number,
  env: Env,
  cache: KVNamespace
): Promise<DefensivePlayer[]> {
  const cacheKey = `tank01_def_stats_${season}`;

  // Check cache first
  const cached = await cache.get(cacheKey);
  if (cached) {
    return JSON.parse(cached);
  }

  if (!isConfigured(env)) {
    throw new Error('Tank01 API not configured. Set RAPIDAPI_KEY secret.');
  }

  // Fetch all players
  const playersRaw = await fetchAllPlayers(env, cache);
  const result: DefensivePlayer[] = [];

  for (const playerData of playersRaw) {
    const pos = String(playerData.pos || '').toUpperCase();

    // Only include defensive positions
    if (!DEFENSIVE_POSITIONS.includes(pos)) {
      continue;
    }

    const fantasyPos = POSITION_MAP[pos] || 'LB';
    const defStats = playerData.stats?.Defense || {};

    const tackles = safeInt(defStats.totalTackles);
    const solo = safeInt(defStats.soloTackles);
    const sacks = safeFloat(defStats.sacks);
    const tfl = safeInt(defStats.tfl);
    const qbHits = safeInt(defStats.qbHits);
    const pd = safeInt(defStats.passDeflections);
    const ints = safeInt(defStats.defensiveInterceptions);
    const ff = safeInt(defStats.forcedFumbles);
    const fr = safeInt(defStats.fumblesRecovered);
    const defTd = safeInt(defStats.defTD);

    // Skip players without relevant stats
    if (tackles === 0 && sacks === 0 && ints === 0) {
      continue;
    }

    result.push({
      id: String(playerData.playerID || playerData.espnID || ''),
      name: playerData.longName || playerData.espnName || 'Unknown',
      team: playerData.team || '',
      teamAbbr: playerData.team || '',
      photoUrl: playerData.espnHeadshot || '',
      espnPosition: pos,
      fantasyPosition: fantasyPos,
      age: playerData.age ? safeInt(playerData.age) : null,
      experience: playerData.exp ? safeInt(playerData.exp) : null,
      jerseyNumber: playerData.jerseyNum ? String(playerData.jerseyNum) : null,
      stats: {
        tackles,
        sacks,
        tfl,
        qbHits,
        passesDefended: pd,
        interceptions: ints,
        forcedFumbles: ff,
        soloTackles: solo,
        assistTackles: tackles > solo ? tackles - solo : 0,
        tacklesWithAssist: 0,
        tflYards: 0,
        sackYards: 0,
        intYards: 0,
        defensiveTds: defTd,
        fumbleRecoveryOwn: 0,
        fumbleRecoveryOpp: fr,
        safeties: 0,
      },
    });
  }

  // Sort by tackles descending
  result.sort((a, b) => b.stats.tackles - a.stats.tackles);

  // Cache for 1 hour
  if (result.length > 0) {
    await cache.put(cacheKey, JSON.stringify(result), {
      expirationTtl: 3600,
    });
  }

  return result;
}

/**
 * Get offensive stats from Tank01 API
 */
export async function getOffensiveStats(
  season: number,
  env: Env,
  cache: KVNamespace
): Promise<OffensivePlayer[]> {
  const cacheKey = `tank01_off_stats_${season}`;

  // Check cache first
  const cached = await cache.get(cacheKey);
  if (cached) {
    return JSON.parse(cached);
  }

  if (!isConfigured(env)) {
    throw new Error('Tank01 API not configured. Set RAPIDAPI_KEY secret.');
  }

  const playersRaw = await fetchAllPlayers(env, cache);
  const result: OffensivePlayer[] = [];

  for (const playerData of playersRaw) {
    const pos = String(playerData.pos || '').toUpperCase();

    // Only include offensive positions
    if (!OFFENSIVE_POSITIONS.includes(pos)) {
      continue;
    }

    const passing = playerData.stats?.Passing || {};
    const rushing = playerData.stats?.Rushing || {};
    const receiving = playerData.stats?.Receiving || {};

    // Passing stats
    const completions = safeInt(passing.passCompletions);
    const attempts = safeInt(passing.passAttempts);
    const passingYards = safeInt(passing.passYds);
    const passingTds = safeInt(passing.passTD);
    const ints = safeInt(passing.int);
    const sacksTaken = safeInt(passing.sacked);

    // Rushing stats
    const carries = safeInt(rushing.carries);
    const rushingYards = safeInt(rushing.rushYds);
    const rushingTds = safeInt(rushing.rushTD);

    // Receiving stats
    const targets = safeInt(receiving.targets);
    const receptions = safeInt(receiving.receptions);
    const receivingYards = safeInt(receiving.recYds);
    const receivingTds = safeInt(receiving.recTD);

    // Calculate passer rating
    let passerRating = 0;
    if (attempts > 0) {
      const a = Math.max(0, Math.min(2.375, ((completions / attempts) - 0.3) * 5));
      const b = Math.max(0, Math.min(2.375, ((passingYards / attempts) - 3) * 0.25));
      const c = Math.max(0, Math.min(2.375, (passingTds / attempts) * 20));
      const d = Math.max(0, Math.min(2.375, 2.375 - ((ints / attempts) * 25)));
      passerRating = Math.round(((a + b + c + d) / 6) * 100 * 10) / 10;
    }

    const totalYards = passingYards + rushingYards + receivingYards;
    if (totalYards === 0) {
      continue;
    }

    result.push({
      id: String(playerData.playerID || playerData.espnID || ''),
      name: playerData.longName || playerData.espnName || 'Unknown',
      team: playerData.team || '',
      teamAbbr: playerData.team || '',
      photoUrl: playerData.espnHeadshot || '',
      position: pos,
      age: playerData.age ? safeInt(playerData.age) : null,
      experience: playerData.exp ? safeInt(playerData.exp) : null,
      jerseyNumber: playerData.jerseyNum ? String(playerData.jerseyNum) : null,
      stats: {
        completions,
        attempts,
        passingYards,
        passingTds,
        interceptions: ints,
        passerRating,
        passingAirYards: 0,
        passingYac: 0,
        passingFirstDowns: 0,
        sacks: sacksTaken,
        sackYards: 0,
        carries,
        rushingYards,
        rushingTds,
        rushingFirstDowns: 0,
        targets,
        receptions,
        receivingYards,
        receivingTds,
        receivingAirYards: 0,
        receivingYac: 0,
        receivingFirstDowns: 0,
        fumbles: 0,
        fumblesLost: 0,
        targetShare: 0,
        airYardsShare: 0,
        wopr: 0,
        racr: 0,
        pacr: 0,
        fantasyPoints: 0,
        fantasyPointsPpr: 0,
      },
    });
  }

  // Sort by total yards descending
  result.sort((a, b) => {
    const aTotal = a.stats.passingYards + a.stats.rushingYards + a.stats.receivingYards;
    const bTotal = b.stats.passingYards + b.stats.rushingYards + b.stats.receivingYards;
    return bTotal - aTotal;
  });

  // Cache for 1 hour
  if (result.length > 0) {
    await cache.put(cacheKey, JSON.stringify(result), {
      expirationTtl: 3600,
    });
  }

  return result;
}
